### PR TITLE
[feature] add config variable to perturb CH4 boundary conditions

### PR DESCRIPTION
**Pull request by nicholasbalasus**
### Name and Institution (Required)

Name: Nick Balasus
Institution: Harvard University

### Confirm you have reviewed the following documentation

- [x] [Contributing guidelines](https://geos-chem.readthedocs.io/en/stable/reference/CONTRIBUTING.html)

### Describe the update

Adds a variable to `gesochem_config.yml.CH4` allowing for the perturbation of boundary conditions edges by a certain number of ppb.

### Expected changes

No changes to any simulations.

### Reference(s)

n/a

### Related Github Issue(s)

n/a

### CODE REVIEW

**Comment by nicholasbalasus** [comment]
Testing with this bash script and this python script:
[run-bash.txt](https://github.com/geoschem/geos-chem/files/12498456/run-bash.txt)
[run-python.txt](https://github.com/geoschem/geos-chem/files/12498458/run-python.txt)

Text results:
```
no_feature_no_perturb_0N0S0E0W
North boundary enhancement -> 0.00 ppb
South boundary enhancement -> 0.00 ppb
East boundary enhancement -> 0.00 ppb
West boundary enhancement -> 0.00 ppb

yes_feature_no_perturb_0N0S0E0W
North boundary enhancement -> 0.00 ppb
South boundary enhancement -> 0.00 ppb
East boundary enhancement -> 0.00 ppb
West boundary enhancement -> 0.00 ppb

yes_feature_yes_perturb_0N0S0E0W
North boundary enhancement -> 0.00 ppb
South boundary enhancement -> 0.00 ppb
East boundary enhancement -> 0.00 ppb
West boundary enhancement -> 0.00 ppb

yes_feature_yes_perturb_0N0S0E10W
North boundary enhancement -> 0.00 ppb
South boundary enhancement -> 0.00 ppb
East boundary enhancement -> 0.00 ppb
West boundary enhancement -> 10.00 ppb

yes_feature_yes_perturb_0N0S10E0W
North boundary enhancement -> 0.00 ppb
South boundary enhancement -> 0.00 ppb
East boundary enhancement -> 10.00 ppb
West boundary enhancement -> 0.00 ppb

yes_feature_yes_perturb_0N10S0E0W
North boundary enhancement -> 0.00 ppb
South boundary enhancement -> 10.00 ppb
East boundary enhancement -> 0.00 ppb
West boundary enhancement -> 0.00 ppb

yes_feature_yes_perturb_10N0S0E0W
North boundary enhancement -> 10.00 ppb
South boundary enhancement -> 0.00 ppb
East boundary enhancement -> 0.00 ppb
West boundary enhancement -> 0.00 ppb

yes_feature_yes_perturb_50N50S50E50W
North boundary enhancement -> 50.00 ppb
South boundary enhancement -> 50.00 ppb
East boundary enhancement -> 50.00 ppb
West boundary enhancement -> 50.00 ppb
```

Plot results:
image

**Comment by nicholasbalasus** [comment]
Let me know what you think @msulprizio. It might make more sense to put everything I put in `hco_utilities_gc_mod.F90` into `set_boundary_conditions_mod.F90` where the same structure of for loops exists, but I opted to do it like this so that `State_Chm%BoundaryCond` was edited, not `State_Chm%Species`. It shouldn't matter for the output, but maybe for cleanliness?

**Comment by jimmielin** [comment]
Hi @nicholasbalasus, this looks great, thanks for working on this.

My two cents is that we could have this code in `set_boundary_conditions_mod.F90` right above where the boundary conditions are actually set. I think this will not impact the output of the boundary conditions in diagnostics; the `State_Chm%BoundaryCond` will still output the modified value, because `set_boundary_conditions` runs before History output. This way we could reuse the loop and keep GEOS-Chem-only code away from the HEMCO + GEOS-Chem code in `hco_utilities_gc_mod.F90`.

I would also suggest adding a check for `Input_Opt%ITS_A_CH4_SIM` so it is explicit that this code is meant for the CH4 simulation.

**Comment by nicholasbalasus** [comment]
Thanks, @jimmielin! I agree with you and will add these changes.

**Comment by nicholasbalasus** [comment]
Hi @jimmielin - I moved the code to `set_boundary_conditions_mod.F90` in a way that should be identical, but now the boundary conditions are being increased by a factor of 23.348 too much. For example, when I specify 10 ppb in the config file, they increase by 233.48 ppb. When I specify 50 ppb, they increase by 1167.39 ppb. I've printed to make sure `MW_g_CH4` is 16.04 and `AIRMW` is 28.964. Does that factor of 23.348 make you think of anything that might be going wrong?

**Comment by jimmielin** [comment]
Hi @nicholasbalasus, this is very strange. I looked at the code closely but couldn't find why it would have an extra factor. 

Some debugging ideas:
1) Could you confirm if the extra 23.348 shows up in both SpeciesConcVV_CH4 and the Chem_BoundaryCond_CH4 diagnostics in a run that sets 10.0 in all sides and the four corner grid boxes appear to have 10 ppb * 23.348 * 2 in them?
2) Could try writing out `State_Chm%BoundaryCond` value at select grid boxes (e.g. I=2 L=2...) before and after adding 10.0 ppb and see if data is in expected units consistent with the BC input netCDF. If it turns out we are using the wrong units in the file instead of the mol/mol expected then it'll be something really surprising...
3) Or I wonder if it is a parallelization issue. Could try
```
srun -p test -N 1 -n 1 --pty --mem 30960 -t 0-06:00 /bin/bash
```
and running `OMP_NUM_THREADS=1 ./gcclassic | tee log.log` in interactive (it'll be a little slow) to see if it still the same?

I can look into this more closely tomorrow.

**Comment by nicholasbalasus** [comment]
Thanks, @jimmielin! Thankfully it looks to be a little more simple and I think it is actually related to time steps. That is, because we are doing this in `set_boundary_conditions_mod.F90`, every time I "set" the boundary conditions this way, they're being increased by 10 ppb (for example) until they are reset at 3 hours again. For example...

```
yes_feature_yes_perturb_0N0S10E0W
t = 0
North boundary enhancement -> 0.00 ppb
South boundary enhancement -> 0.00 ppb
East boundary enhancement -> 110.00 ppb
West boundary enhancement -> 0.00 ppb
t = 1
North boundary enhancement -> 0.00 ppb
South boundary enhancement -> 0.00 ppb
East boundary enhancement -> 230.00 ppb
West boundary enhancement -> 0.00 ppb
t = 2
North boundary enhancement -> 0.00 ppb
South boundary enhancement -> 0.00 ppb
East boundary enhancement -> 350.00 ppb
West boundary enhancement -> 0.00 ppb
t = 3
North boundary enhancement -> 0.00 ppb
South boundary enhancement -> 0.00 ppb
East boundary enhancement -> 120.00 ppb
West boundary enhancement -> 0.00 ppb
t = 4
North boundary enhancement -> 0.00 ppb
South boundary enhancement -> 0.00 ppb
East boundary enhancement -> 240.00 ppb
West boundary enhancement -> 0.00 ppb
t = 5
North boundary enhancement -> 0.00 ppb
South boundary enhancement -> 0.00 ppb
East boundary enhancement -> 360.00 ppb
West boundary enhancement -> 0.00 ppb
```

To handle this, I think `set_boundary_conditions.F90` would need to be aware of if `State_Chm%BoundaryCond(I,J,L,N)` has already been altered. Alternatively, we could just modify `State_Chm%Species(N)%Conc(I,J,L)`. What do you think?

**Comment by yantosca** [comment]
@jimmielin @nicholasbalasus: You might want to check the value of `State_Chm%Spc_Units` in the new location.  If the units are different then that may explain it.

**Comment by nicholasbalasus** [comment]
Thanks @yantosca! Haipeng added a check for the units 

```
   ! Ensure species array is in kg/kg as State_Chm%BoundaryCond is in kg/kg dry
   IF ( TRIM(State_Chm%Spc_Units) /= 'kg/kg dry' ) THEN
      IF ( Input_Opt%amIRoot ) THEN
          WRITE(6,*) 'Unit check failure: Current units are ', &
               State_Chm%Spc_Units, ', expected kg/kg dry'
      ENDIF
      CALL GC_Error( 'Unit check failure: Cannot apply nested-grid boundary conditions if units are not kg/kg dry. Your run may have failed previous to this error.', RC, LOC )
      RETURN
   ENDIF
```

so we should be all good here... I think the "continuous" instead of "once" perturbation is my issue.

**Comment by jimmielin** [comment]
> To handle this, I think set_boundary_conditions.F90 would need to be aware of if State_Chm%BoundaryCond(I,J,L,N) has already been altered. Alternatively, we could just modify State_Chm%Species(N)%Conc(I,J,L). What do you think?

Hi @nicholasbalasus, thanks for figuring this out! I think in order to be able to 'see' this perturbation in the diagnostic output, I would suggest modifying BoundaryCond only once (with a logical switch) and avoid modifying Species%Conc directly.

I think we could do it like this. Adding a logical somewhere in `State_Chm` (e.g. `State_Chm%Is_CH4BCPerturbed = .false.`) and set it to `.false.` from `GET_BOUNDARY_CONDITIONS` every time BCs are read in; then set it to `.true.` when `SET_BOUNDARY_CONDITIONS` modifies it. But maybe it is too much work, feel free to implement it any other way.

Thanks!

**Comment by nicholasbalasus** [comment]
Thanks, @jimmielin! This is now working! Including the plot again because I plotted the left column incorrectly above.
image

**Comment by nicholasbalasus** [comment]
@yantosca this is fixed now. Do you think my definition of `State_Chm%IsCH4BCPerturbed` is okay?

File: run/GCClassic/geoschem_config.yml.templates/geoschem_config.yml.CH4 [hunk 1]
```diff
@@ -83,6 +83,8 @@ CH4_simulation_options:
     state_vector_element_number: 0
     use_emission_scale_factor: false
     use_OH_scale_factors: false
+    perturb_CH4_boundary_conditions: false
+    CH4_boundary_condition_ppb_increase_NSEW: [0.0, 0.0, 0.0, 0.0]
```
Comment (msulprizio) [inline run/GCClassic/geoschem_config.yml.templates/geoschem_config.yml.CH4:87]: This line should also be added to the geoschem_config.yml files for tagCH4 and carbon simulations.

File: GeosCore/hco_utilities_gc_mod.F90 [hunk 2]
```diff
@@ -2573,6 +2573,12 @@ SUBROUTINE Get_Boundary_Conditions( Input_Opt, State_Chm, State_Grid, &
       ! Free pointer
       SpcInfo => NULL()
 
+      ! Each time the boundary conditions are read, they have no longer been perturbed
+      ! This value is sometimes set to True in set_boundary_conditions_mod.F90
+      IF (Input_Opt%ITS_A_CH4_SIM .AND. Input_Opt%PerturbCH4BoundaryConditions ) THEN
```
Comment (msulprizio) [inline GeosCore/hco_utilities_gc_mod.F90:2578]: I recommend adding `.OR. Input_OPt%ITS_A_CARBON_SIM` here (enclosed in parentheses with ITS_A_CH4_SIM)

File: GeosCore/set_boundary_conditions_mod.F90 [hunk 3]
```diff
@@ -133,16 +136,33 @@ SUBROUTINE Set_Boundary_Conditions( Input_Opt, State_Chm, State_Grid, RC )
       ! Get the species ID from the advected species ID
       N = State_Chm%Map_Advect(NA)
 
+      ! Optionally perturb the CH4 boundary conditions
+      ! Use ppb values specified in geoschem_config.yml 
+      ! Convert to [kg/kg dry] (nbalasus, 8/31/2023)
+      Perturb_CH4_BC = ( State_Chm%SpcData(N)%Info%Name == "CH4" .AND. &
+                         Input_Opt%ITS_A_CH4_SIM .AND. &
```
Comment (msulprizio) [inline GeosCore/set_boundary_conditions_mod.F90:143]: I recommend adding `.OR. Input_Opt%ITS_A_CARBON_SIM`

**Review by msulprizio — CHANGES_REQUESTED** [review]
A few changes need to be made to allow integrations to pass (for tagCH4 and carbon simulations). Otherwise, these updates look good.

**Comment by nicholasbalasus** [comment]
Thanks @msulprizio! Testing now for all three simulation types.

**Review by msulprizio — APPROVED** [review]

File: Headers/state_chm_mod.F90 [hunk 4]
```diff
@@ -603,6 +604,9 @@ SUBROUTINE Zero_State_Chm( State_Chm, RC )
     State_Chm%Do_SulfateMod_Cld     = .FALSE.
     State_Chm%Do_SulfateMod_SeaSalt = .FALSE.
 
+   ! Flag if CH4 BC has been perturbed or not
```
Comment (yantosca) [inline Headers/state_chm_mod.F90:607]: @nicholasbalasus: this looks OK to me, to answer your question.

**Comment by yantosca** [comment]
Integration tests are running

**Comment by yantosca** [comment]
All GEOS-Chem Classic integration tests passed (except for TOMAS, which is a known issue):
==============================================================================
GEOS-Chem Classic: Execution Test Results

GCClassic #cee8090 Merge PR #1963 (Add fixes for single-species carbon simulation)
GEOS-Chem #623e65097 Merge PR #1934 (Add config variable to perturb CH4 BCs)
HEMCO     #daf54d0 Merge PR #235 (Bug fix for vertical regridding)

Using 24 OpenMP threads
Number of execution tests: 27

Submitted as SLURM job: 2666166
==============================================================================
 
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 25
Execution tests failed: 2
Execution tests not yet completed: 0
```
Furthermore, all integration tests were zero-diff w/r/t the previous integration test (PR #1963) except for simulations with known issues:
```console
Checking gc_4x5_merra2_fullchem_APM
   -> 2 differences found in OutputDir
      * GCC_r25/rundirs/gc_4x5_merra2_fullchem_APM/OutputDir/GEOSChem.Metrics.20190701_0000z.nc4 
        GCC_r26/rundirs/gc_4x5_merra2_fullchem_APM/OutputDir/GEOSChem.Metrics.20190701_0000z.nc4 
      * GCC_r25/rundirs/gc_4x5_merra2_fullchem_APM/OutputDir/GEOSChem.SpeciesConc.20190701_0000z.nc4 
        GCC_r26/rundirs/gc_4x5_merra2_fullchem_APM/OutputDir/GEOSChem.SpeciesConc.20190701_0000z.nc4 
   -> 1 difference found in Restarts
      * GCC_r25/rundirs/gc_4x5_merra2_fullchem_APM/Restarts/GEOSChem.Restart.20190701_0100z.nc4 
        GCC_r26/rundirs/gc_4x5_merra2_fullchem_APM/Restarts/GEOSChem.Restart.20190701_0100z.nc4 

Checking gc_4x5_merra2_fullchem_RRTMG
   -> 1 difference found in OutputDir
      * GCC_r25/rundirs/gc_4x5_merra2_fullchem_RRTMG/OutputDir/GEOSChem.RRTMG.20190701_0000z.nc4 
        GCC_r26/rundirs/gc_4x5_merra2_fullchem_RRTMG/OutputDir/GEOSChem.RRTMG.20190701_0000z.nc4 
   -> No differences in Restarts
```

**Comment by nicholasbalasus** [comment]
Thanks @msulprizio @yantosca @jimmielin !

**Comment by yantosca** [comment]
All GCHP integration tests passed:
```console
==============================================================================
GCHP: Execution Test Results

GCClassic #d798220 Merge PR #1963 (Add fixes for single-species carbon simulation)
GEOS-Chem #623e65097 Merge PR #1934 (Add config variable to perturb CH4 BCs)
HEMCO     #daf54d0 Merge PR #235 (Bug fix for vertical regridding)

Number of execution tests: 5

Submitted as SLURM job: 2668528
==============================================================================
 
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 5
Execution tests failed: 0
Execution tests not yet completed: 0

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
Also all integration tests were zero-diff w/r/t the prior integration test (PR #1963).
